### PR TITLE
Allow for heterogenous List and Dict values + Improve container typing algorithm

### PR DESCRIPTION
**Pull request by ansley**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#57137**

Differential Revision: [D28231839](https://our.internmc.facebook.com/intern/diff/D28231839)

This PR corrects and expands our typing algorithm for unannotated, non-empty dicts and lists. Previously, to verify type correctness for an unannotated, non-empty container, we had gotten the type of the first element in the container, then checked if each following element was a subtype of the first type. That's too restrictive--what if the first element were a subtype of the second element? Instead, we should type the container by getting the smallest common supertype of all the given elements.

We need slightly different rules for keys and values in dicts, though: because the set of key types is restricted, finding two key types that cannot be unified should cause an error. On the other hand, the set of value types is not restricted, so we should be able to use `Any` as a valid supertype. We need to keep the set of keys restricted since the keys are used to generate and match schemas.

This does not break backwards compatibility, because the default element type is the smallest supertype of all the given types. So, if someone creates an unannotated dict where the keys are all `str` and the values are all `torch.Tensor`, the dict will be inferred to `Dict[str, Tensor]` just like it was before. Empty lists are still typed as `List[torch.Tensor]`, and empty dicts are still typed as `Dict[str, Tensor]`.

This PR unblocks three engineers on an FB-internal team and improves FX-TorchScript compatibility.

### CODE REVIEW

File: test/jit/test_typing.py [hunk 1]
```diff
@@ -73,11 +73,17 @@ def test_dict_tensor_key(a, t):
...
+        def fn(x):
+            d = {"foo": torch.tensor(2),
+                 "bar": {"23": torch.tensor(3)}}
+            d["baz"] = x
+            t = d["foo"]
+            if isinstance(t, torch.Tensor):
+                d["bar"] = torch.add(t, t)
+            return d
+
+        self.checkScript(fn, (torch.rand(2, 3),))
```
Comment (gmagogsfm) [inline test/jit/test_typing.py:86]: Optional:

Use filecheck to verify type is correct in IR

File: test/jit/test_typing.py [hunk 2]
```diff
@@ -73,11 +73,17 @@ def test_dict_tensor_key(a, t):
         self.checkScript(test_dict_tensor_key, (dict_a, inp1))
         self.checkScript(test_dict_tensor_key, (dict_a, inp2))
 
-    def test_dict_types(self):
-        with self.assertRaisesRegex(RuntimeError, "single type"):
-            @torch.jit.script
-            def foo():
-                new_item = {'score': [1.0], 'ys': [1, 2, 3]}
+    def test_dict_value_type_refinement_defaults_to_Any(self):
```
Comment (gmagogsfm) [inline test/jit/test_typing.py:76]: Are other ways of constructing heterogenous-value dicts supported too? like from dict comprehension, from gradually inserting different types of values etc.

File: torch/csrc/jit/frontend/ir_emitter.cpp [hunk 3]
```diff
@@ -3675,24 +3675,25 @@ struct to_ir {
...
-        };
-        checkTypeOfValues(key_type, "keys", keys, key_trees);
-        checkTypeOfValues(value_type, "values", values, value_trees);
+        }
+
+        for (size_t i = 0, N = values.size(); i < N; ++i) {
+          std::stringstream ss;
+          if (!values[i]->type()->isSubtypeOfExt(value_type, &ss) &&
+              !value_type->isSubtypeOfExt(values[i]->type(), &ss)) {
+            value_type = AnyType::get();
```
Comment (gmagogsfm) [inline torch/csrc/jit/frontend/ir_emitter.cpp:3694]: In case of `value_type` is a subtype of type of following values, should we update `value_type` to the super type?

Additionally, it looks like what you need is `unifyTypes` here, which would naturally return `Any` if types do not have subtyping relationship.

File: torch/csrc/jit/frontend/ir_emitter.cpp [hunk 4]
```diff
@@ -3675,24 +3675,25 @@ struct to_ir {
...
-                  << "Dict " << what
-                  << " must contain only a single type, expected: "
-                  << type->repr_str() << " but found "
-                  << values[i]->type()->repr_str() << " instead.\n"
-                  << ss.str();
-            }
+        for (size_t i = 0, N = values.size(); i < N; ++i) {
+          std::stringstream ss;
+          if (!keys[i]->type()->isSubtypeOfExt(key_type, &ss) &&
+              !key_type->isSubtypeOfExt(keys[i]->type(), &ss)) {
```
Comment (gmagogsfm) [inline torch/csrc/jit/frontend/ir_emitter.cpp:3681]: Same here, should `key_type` be updated to super type?

Comment (ansley) [reply torch/csrc/jit/frontend/ir_emitter.cpp:3681]: This might be a good follow-up project, but it's a lot bigger than it seems on the surface. The reason we restrict the allowable keys to a small set of types is that we generate a unique schema for each type

**Comment by ansley** [comment]
@ansley has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28231839).

File: torch/csrc/jit/frontend/ir_emitter.cpp [hunk 5]
```diff
@@ -1320,13 +1320,41 @@ struct to_ir {
...
+
+      // Is the current element type the same type or a supertype of the
+      // old element type?
+      bool use_new = lt->getElementType()->isSubtypeOf(out->type());
+      // Is the old element type the same type or a supertype of the
+      // current element type?
+      bool use_old = out->type()->isSubtypeOf(lt->getElementType());
+
+      // If we have `use_old && use_new`, we just keep the old element
+      // type. Otherwise...
```
Comment (gmagogsfm) [inline torch/csrc/jit/frontend/ir_emitter.cpp:1349]: Is this comment incomplete?

File: test/jit/test_typing.py (hunk 1 above)
Comment (ansley) [reply test/jit/test_typing.py:152]: Good suggestion, added

File: torch/csrc/jit/frontend/ir_emitter.cpp (hunk 3 above)
Comment (ansley) [reply torch/csrc/jit/frontend/ir_emitter.cpp]: Fixed, good call

File: test/jit/test_typing.py (hunk 2 above)
Comment (ansley) [reply test/jit/test_typing.py]: I added some new logic and test cases. Please take a look at the test cases I commented out and let me know whether you think we should support that too

File: torch/csrc/jit/frontend/ir_emitter.cpp [hunk 6]
```diff
@@ -3533,6 +3593,47 @@ struct to_ir {
             ->call(tree->range(), method, named_values, {}, 0));
   }
 
+  Value* emitListLiteral(ListLiteral ll, TypePtr type_hint) {
+    auto values = getValues(ll.inputs(), /*maybe_unpack=*/true);
+
+    // determine the element type of the list
+    // if we have a type hint of List[T], use T
+    // if the list is non-empty use type_of(list[0])
+    // otherwise assume it is List[Tensor]
```
Comment (gmagogsfm) [inline torch/csrc/jit/frontend/ir_emitter.cpp:3602]: This comment is no longer accurate

File: torch/csrc/jit/frontend/ir_emitter.cpp (hunk 5 above)
Comment (ansley) [reply torch/csrc/jit/frontend/ir_emitter.cpp:1349]: Thanks for pointing out that this comment is unclear. The "Otherwise..." meant "Otherwise, we do the logic directly below this comment". I'll change it!

File: torch/csrc/jit/frontend/ir_emitter.cpp [hunk 7]
```diff
@@ -1320,13 +1320,41 @@ struct to_ir {
...
+      bool use_old = out->type()->isSubtypeOf(lt->getElementType());
+
+      // If we have `use_old && use_new`, we just keep the old element
+      // type. Otherwise...
+      if (!use_old && use_new) {
+        list_value->setType(ListType::create(out->type()));
+      } else if (!use_old && !use_new) {
+        list_value->setType(ListType::create(AnyType::get()));
+      }
+
```
[the remark this answers is not in the document]
Comment (ansley) [reply torch/csrc/jit/frontend/ir_emitter.cpp:1355]: Good point

File: torch/csrc/jit/frontend/ir_emitter.cpp (hunk 6 above)
Comment (ansley) [reply torch/csrc/jit/frontend/ir_emitter.cpp:3602]: Thanks!

File: torch/csrc/jit/frontend/ir_emitter.cpp [hunk 8]
```diff
@@ -3689,24 +3751,23 @@ struct to_ir {
...
-              throw ErrorReport(trees[i])
-                  << "Dict " << what
-                  << " must contain only a single type, expected: "
-                  << type->repr_str() << " but found "
-                  << values[i]->type()->repr_str() << " instead.\n"
-                  << ss.str();
-            }
+        for (size_t i = 0, N = values.size(); i < N; ++i) {
+          std::stringstream ss;
+          if (!unifyTypes(keys[i]->type(), key_type, /*default_to_any=*/false)) {
```
[the remark this answers is not in the document]
Comment (ansley) [reply torch/csrc/jit/frontend/ir_emitter.cpp:3756]: Good catch, thanks!

File: test/HowToWriteTestsUsingFileCheck.md [hunk 9]
```diff
@@ -79,6 +79,9 @@ annotations from the example above one would write:
 
 * `CHECK: <pattern>`
   Scans the input until `PATTERN` is found. Fails if the pattern is not found.
+* `CHECK-NEXT: <pattern>`
+  Scans the input on the line immediately following the previous CHECK until
+  `PATTERN` is found. Fails if the pattern is not found on that line.
```
Comment (gmagogsfm) [inline test/HowToWriteTestsUsingFileCheck.md:84]: Consider making separate PR for changes that are not relevant to the functionality of PR

**Review by gmagogsfm — CHANGES_REQUESTED** [review]
There seems to be a few real test failures and lint warnings, could you push another version with them fixed?

File: test/HowToWriteTestsUsingFileCheck.md (hunk 9 above)
Comment (ansley) [reply test/HowToWriteTestsUsingFileCheck.md:84]: Gotcha, will do that next time

File: test/jit/test_typing.py (hunk 2 above)
Comment (ansley) [reply test/jit/test_typing.py]: cc @gmagogsfm Reminder to look at those commented-out test cases :)

File: torch/csrc/jit/frontend/ir_emitter.cpp [hunk 10]
```diff
@@ -1320,13 +1320,31 @@ struct to_ir {
...
     WithInsertPoint guard(comprehension_block);
     auto emit_body = [&]() {
-      auto comprehension_out = emitExpr(lc.elt());
+      Value* out = emitExpr(lc.elt());
+
+      // Make sure that any element types are subtypes of the annotated
+      // type
+      std::stringstream ss;
+      if (type_hint && !out->type()->isSubtypeOfExt(type_hint->expect<ListType>()->getElementType(), &ss)) {
+        throw ErrorReport(loc) << ss.str() << "List type annotation did"
```
Comment (gmagogsfm) [inline torch/csrc/jit/frontend/ir_emitter.cpp:1329]: `ss` contains the reason for not satisfying subtype relationship, it should ideally be at the end of error message.

File: torch/csrc/jit/frontend/ir_emitter.cpp [hunk 11]
```diff
@@ -1320,13 +1320,31 @@ struct to_ir {
...
+                               << " not match the types of the actual "
+                               << "list items";
+      }
+
+      // If we didn't have a type annotation, let the type begin as the
+      // first element in the list
       if (!type_set) {
```
Comment (gmagogsfm) [inline torch/csrc/jit/frontend/ir_emitter.cpp:1336]: Why is this conditional type inference needed when the next step is to infer unified type unconditionally?

File: torch/csrc/jit/frontend/ir_emitter.cpp (hunk 10 above)
Comment (ansley) [reply torch/csrc/jit/frontend/ir_emitter.cpp:1329]: thanks!

File: torch/csrc/jit/frontend/ir_emitter.cpp (hunk 11 above)
Comment (ansley) [reply torch/csrc/jit/frontend/ir_emitter.cpp:1336]: If we didn't have a type annotation, the type of the list would be set to `Tensor`. We don't want to unify this default type with the actual elements in the list. I've updated the comment to explain this edge case

File: torch/csrc/jit/frontend/ir_emitter.cpp [hunk 12]
```diff
@@ -3689,24 +3743,26 @@ struct to_ir {
...
-        checkTypeOfValues(key_type, "keys", keys, key_trees);
-        checkTypeOfValues(value_type, "values", values, value_trees);
+        }
+
+        if (!values.empty()) {
+          value_type = std::accumulate(
+              values.begin(),
+              values.end(),
+              value_type,
+              [&](TypePtr a, Value* b) {
```
[the remark this answers is not in the document]
Comment (ansley) [reply torch/csrc/jit/frontend/ir_emitter.cpp:3762]: Good catch! Please take a look at my new algorithm

File: torch/csrc/jit/frontend/ir_emitter.cpp [hunk 13]
```diff
@@ -3533,6 +3573,59 @@ struct to_ir {
...
+      // the actual list values, so this condition is left as a warning
+      // for backwards compatibility
+      if (!(*rhs_type)->isSubtypeOf(elem_type)) {
+        TORCH_WARN("List annotation ", type_hint->repr_str(), " did "
+        "not match the types of the given list elements, which were "
+        "unified to ", (*rhs_type)->repr_str());
+      }
+
+      if (*rhs_type == AnyType::get()) {
+        TORCH_WARN(
```
[the remark this answers is not in the document]
Comment (ansley) [reply torch/csrc/jit/frontend/ir_emitter.cpp:3625]: Done

**Comment by ansley** [comment]
@ansley has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28231839).

File: torch/csrc/jit/frontend/ir_emitter.cpp [hunk 14]
```diff
@@ -3689,24 +3754,79 @@ struct to_ir {
...
+          auto get_greater_type =
+              [](const TypePtr t1, const TypePtr t2) -> c10::optional<TypePtr> {
+            if (t1->isSubtypeOf(t2)) {
+              return t2;
+            } else if (t2->isSubtypeOf(t1)) {
+              return t1;
+            } else {
+              return c10::nullopt;
+            }
+          };
```
Comment (gmagogsfm) [inline torch/csrc/jit/frontend/ir_emitter.cpp:3789]: This looks very similar to `unifyTypes(t1, t2, /*default_to_any=*false/)`, could you clarify the differences?

File: torch/csrc/jit/frontend/ir_emitter.cpp [hunk 15]
```diff
@@ -3806,8 +3926,8 @@ struct to_ir {
     return emitBuiltinCall(loc, *graph, aten::slice, args, {step_nv});
   }
 
-  // Desugars slice indexing: tensor[begin:end] -> tensor.slice(dim, begin, end,
-  // 1)
+  // Desugars slice indexing: tensor[begin:end] -> tensor.slice(dim, begin,
```
Comment (gmagogsfm) [inline torch/csrc/jit/frontend/ir_emitter.cpp:3929]: Changes that start from here to end of file seem to be lint only, could you move them to another PR?

File: torch/csrc/jit/frontend/ir_emitter.cpp [hunk 16]
```diff
@@ -3689,24 +3754,79 @@ struct to_ir {
...
+            }
+            if (!added) {
+              seen.push_back(type);
             }
           }
-        };
-        checkTypeOfValues(key_type, "keys", keys, key_trees);
-        checkTypeOfValues(value_type, "values", values, value_trees);
+
+          TypePtr inferred_type = seen.size() == 1 ? seen[0] : AnyType::get();
```
Comment (gmagogsfm) [inline torch/csrc/jit/frontend/ir_emitter.cpp:3813]: Could you clarify how this condition works for the motivating use case on line 3773?

It looks like that, at end of algorithm, `seen` would be [T_1, T_1], which would fail check of `seen.size() == 1`, making the final type `Any`. If this is indeed the case, we can change type of `seen` to `std::unordered_set`, which fixes this issue by automatically deduping elements. Additionally, `unordered_set` has faster look up time, which makes it easier to check if a `type` has been visited before.

File: torch/csrc/jit/frontend/ir_emitter.cpp [hunk 17]
```diff
@@ -3689,24 +3754,79 @@ struct to_ir {
...
+                << "only a single type, expected: " << key_type->repr_str()
+                << " but found " << keys[i]->type()->repr_str() << " instead.\n"
+                << ss.str();
+          }
+        }
+
+        if (!values.empty()) {
+          // Get the smallest common supertype of all the values. We
+          // need to use the O(N^2) algorithm instead of using
+          // `std::accumulate` to account for the case that two types
```
Comment (gmagogsfm) [inline torch/csrc/jit/frontend/ir_emitter.cpp:3771]:
```suggestion
          // need to use the O(N^2) algorithm to account for the case that two types
```

File: torch/csrc/jit/frontend/ir_emitter.cpp [hunk 18]
```diff
@@ -3689,24 +3754,79 @@ struct to_ir {
...
+          TypePtr inferred_type = seen.size() == 1 ? seen[0] : AnyType::get();
+
+          std::stringstream ss;
+          TORCH_CHECK(
+              !type_hint ||
+                  inferred_type->isSubtypeOf(
+                      type_hint->expect<DictType>()->getValueType()),
+              "Type hint for dict values was ",
+              type_hint->repr_str(),
+              " but the greatest common supertype of the given values",
```
Comment (gmagogsfm) [inline torch/csrc/jit/frontend/ir_emitter.cpp:3822]: IIUC, this can only happen when there is at least one value that is not a sub-type of type hint. Is this accurate? If so, maybe we could make the error message simpler and more actionable if we loop through all value types and check their type relationship with type hint. That way, we would be able to tell user which exactly value offends typing rules, instead of erroring on an inferred supertype that user may not have specified themselves.

File: torch/csrc/jit/frontend/ir_emitter.cpp [hunk 19]
```diff
@@ -3689,24 +3754,79 @@ struct to_ir {
...
-          for (size_t i = 0, N = values.size(); i < N; ++i) {
-            std::stringstream ss;
-            if (!values[i]->type()->isSubtypeOfExt(type, &ss)) {
-              throw ErrorReport(trees[i])
-                  << "Dict " << what
-                  << " must contain only a single type, expected: "
-                  << type->repr_str() << " but found "
-                  << values[i]->type()->repr_str() << " instead.\n"
-                  << ss.str();
+        for (size_t i = 0, N = keys.size(); i < N; ++i) {
```
Comment (gmagogsfm) [inline torch/csrc/jit/frontend/ir_emitter.cpp:3757]:
```suggestion
        for (size_t i = 0, n = keys.size(); i < n; ++i) {
```

**Review by gmagogsfm — CHANGES_REQUESTED** [review]

File: torch/csrc/jit/frontend/ir_emitter.cpp [hunk 20]
```diff
@@ -1320,13 +1320,35 @@ struct to_ir {
...
       if (!type_set) {
-        list_value->setType(ListType::create(comprehension_out->type()));
+        list_value->setType(ListType::create(out->type()));
         type_set = true;
       }
+
+      ListTypePtr lt = list_value->type()->expect<ListType>();
+      auto unified = unifyTypes(
+          lt->getElementType(), out->type(), /*default_to_any=*/true);
+      list_value->setType(ListType::create(*unified));
```
[the remark this answers is not in the document]
Comment (ansley) [reply torch/csrc/jit/frontend/ir_emitter.cpp:1348]: Good point

File: test/jit/test_typing.py [hunk 21]
```diff
@@ -73,11 +74,80 @@ def test_dict_tensor_key(a, t):
...
+            l = [t for t in l_]    # noqa: C416
+            l.append(tup3)
+            tup4 = l[0]
+            if torch.jit.isinstance(tup4, Tuple[str, torch.Tensor]):
+                t = tup4[1]
+                if isinstance(t, torch.Tensor):
+                    l[0] = (tup4[0], torch.add(t, t))
+            return l
+
+        self.checkScript(fn, (torch.rand(2, 3),))
```
[the remark this answers is not in the document]
Comment (ansley) [reply test/jit/test_typing.py:112]: I had no idea! Thanks!

File: test/jit/test_typing.py [hunk 22]
```diff
@@ -73,11 +74,80 @@ def test_dict_tensor_key(a, t):
...
+
+        graph = torch.jit.script(fn).graph
+
+        FileCheck().check(r"(str, Any)[]").run(graph)
```
[the remark this answers is not in the document]
Comment (ansley) [reply test/jit/test_typing.py:95]: Perhaps this is proof that we should change '(' to `Tuple`... :)

The real type of the list is `List[Tuple[str, Any]]`. I'll change it to be the full `ListConstruct` line!

File: torch/csrc/jit/frontend/ir_emitter.cpp [hunk 23]
```diff
@@ -1320,13 +1320,35 @@ struct to_ir {
...
-      auto comprehension_out = emitExpr(lc.elt());
+      Value* out = emitExpr(lc.elt());
+
+      // Make sure that any element types are subtypes of the annotated
+      // type
+      std::stringstream ss;
+      if (type_hint &&
+          !out->type()->isSubtypeOfExt(
+              type_hint->expect<ListType>()->getElementType(), &ss)) {
+        throw ErrorReport(loc) << "List type annotation did not match "
```
[the remark this answers is not in the document]
Comment (ansley) [reply torch/csrc/jit/frontend/ir_emitter.cpp:1331]: Done! On a side note, I had to allow this behavior in `emitListLiteral` for BC purposes, but I think it really should be an error. How would you feel if I made it an error and simply changed our broken tests? I could stack another PR on top of this as a Better Engineering effort

File: torch/csrc/jit/frontend/ir_emitter.cpp [hunk 24]
```diff
@@ -1365,10 +1387,33 @@ struct to_ir {
...
       auto v = emitExpr(dc.value());
+
+      // Make sure that any key and value types are subtypes of the
+      // annotatated key/value types
+      if (type_hint) {
+        DictTypePtr dict_type_hint = type_hint->expect<DictType>();
+        if (!k->type()->isSubtypeOf(dict_type_hint->getKeyType()) ||
+            !v->type()->isSubtypeOf(dict_type_hint->getValueType())) {
+          throw ErrorReport(loc) << "Dict type annotation did not match "
+                                 << "the types of the actual dict items";
```
[the remark this answers is not in the document]
Comment (ansley) [reply torch/csrc/jit/frontend/ir_emitter.cpp:1398]: All good points, thanks

File: torch/csrc/jit/frontend/ir_emitter.cpp [hunk 25]
```diff
@@ -1365,10 +1387,33 @@ struct to_ir {
...
         type_set = true;
       }
+
+      DictTypePtr dt = dict_value->type()->expect<DictType>();
+
+      c10::optional<TypePtr> unified = unifyTypes(
+          dt->getValueType(),
+          v->type(),
+          /*default_to_any=*/true);
+      dict_value->setType(DictType::create(k->type(), *unified));
```
[the remark this answers is not in the document]
Comment (ansley) [reply torch/csrc/jit/frontend/ir_emitter.cpp:1415]: Done

File: test/jit/test_typing.py (hunk 21 above)
[the remark this answers is not in the document]
Comment (gmagogsfm) [reply test/jit/test_typing.py:112]: No worries.

On a side note, there is no need to hurry in replying to comments, take your time to fix implementation, wait for tests/lint to pass, push to Github and only then reply in comment about what was done to address the issue. Replying early doesn't really add much information. Also, please refrain from resolving comment thread, which can cause an issue to get lost in communication. Ideally only reviewer should resolve comment thread after reaching agreement on the fix.

File: torch/csrc/jit/frontend/ir_emitter.cpp (hunk 15 above)
Comment (ansley) [reply torch/csrc/jit/frontend/ir_emitter.cpp:3929]: Thanks, I'll remove them

File: torch/csrc/jit/frontend/ir_emitter.cpp (hunk 16 above)
Comment (ansley) [reply torch/csrc/jit/frontend/ir_emitter.cpp:3813]: Good point. I'm actually going to open a PR to change `unifyTypeList` to this more forward compatible version (anticipating a future in which we have a less flat type hierarchy)

File: torch/csrc/jit/frontend/ir_emitter.cpp (hunk 23 above)
[the remark this answers is not in the document]
Comment (gmagogsfm) [reply torch/csrc/jit/frontend/ir_emitter.cpp:1331]: It does seem like a real error. Which test is broken?

File: torch/csrc/jit/frontend/ir_emitter.cpp (hunk 16 above)
Comment (gmagogsfm) [reply torch/csrc/jit/frontend/ir_emitter.cpp:3813]: Could you clarify how "changing unityTypeList" can address the incorrect result for motivating use case?

File: test/jit/test_typing.py (hunk 21 above)
[the remark this answers is not in the document]
Comment (ansley) [reply test/jit/test_typing.py:112]: Thanks also for the side note! I'd been using the PR comments as a task list in the past, where a "resolve conversation" meant "done with the task". Moving forward, I'll use PR comments as you're suggesting! It makes a lot of sense!

You and I should talk offline about writing up a "Frontend GitHub Etiquette" doc for new hires. Some of the other reviewers I've worked with during my time at FB have different preferences (e.g. they only want me to silently hit "resolve conversation" unless I disagree with what they said). I think both ways are good, but we should make sure that everyone on the team is on the same page.

File: torch/csrc/jit/frontend/ir_emitter.cpp (hunk 16 above)
Comment (ansley) [reply torch/csrc/jit/frontend/ir_emitter.cpp:3813]: This algorithm is basically doing what `unifyTypeList` is doing; it just accounts for the subtyping relationships you pointed out. Because the number of types in our types system is pretty small, changing `unifyTypeList` from the current `O(N)` algorithm to a forward-compatible `O(N^2)` algorithm would bolster us against the future without causing any real performance degradation. If we introduce polymorphic types in the next year or so and our type system expands greatly, we would need the `O(N^2)` algorithm anyway. So why not go ahead and change it?

We should also make our list emission forwards compatible (which it currently isn't). The new `unifyTypeList` could be used in all four sections of `ir_emitter.cpp` that this PR touches.

**Comment by ansley** [comment]
@ansley has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28231839).

File: torch/csrc/jit/frontend/ir_emitter.cpp (hunk 23 above)
[the remark this answers is not in the document]
Comment (ansley) [reply torch/csrc/jit/frontend/ir_emitter.cpp:1332]: Yes, I deleted [this](https://github.com/pytorch/pytorch/pull/57137/files#diff-377406d8c2c4a9c34da88772f543dc94a0e81753a0fcb6208cdc579a057e627fL574) test case. It seems to me like what the code tests for is a mistake, but this may be me misunderstanding `torch.jit.interface` (there's no documentation for it at all). Please take a look before I ship this

File: torch/csrc/jit/frontend/ir_emitter.cpp (hunk 19 above)
Comment (ansley) [reply torch/csrc/jit/frontend/ir_emitter.cpp]: fixed

File: torch/csrc/jit/frontend/ir_emitter.cpp (hunk 18 above)
Comment (ansley) [reply torch/csrc/jit/frontend/ir_emitter.cpp]: fixed

File: torch/csrc/jit/frontend/ir_emitter.cpp (hunk 17 above)
Comment (ansley) [reply torch/csrc/jit/frontend/ir_emitter.cpp]: fixed

**Comment by ansley** [comment]
@ansley has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28231839).

**Comment by ansley** [comment]
@ansley has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28231839).

**Comment by ansley** [comment]
@ansley has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28231839).

**Comment by ansley** [comment]
@ansley has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28231839).

File: torch/csrc/jit/frontend/ir_emitter.cpp [hunk 26]
```diff
@@ -1366,10 +1415,95 @@ struct to_ir {
...
+              << "`\n"
+              << ss.str();
+        }
+
+        if (!is_key_subtype && !is_value_subtype) {
+          throw ErrorReport(dc)
+              << "Dict type annotation `" << dict_type_hint->repr_str()
+              << "` did not match the types of the "
+              << "actual dict items";
+        }
```
Comment (gmagogsfm) [inline torch/csrc/jit/frontend/ir_emitter.cpp:1455]: This check is superceded by the check of 1457-1459 I think.

File: torch/csrc/jit/frontend/ir_emitter.cpp [hunk 27]
```diff
@@ -1366,10 +1415,95 @@ struct to_ir {
...
+            dt->getValueType()->repr_str(),
+            "\n",
+            dc.range().str());
+      }
+
+      // We only want to set `dict_value` if we don't have a type hint
+      // to allow for the case that `*unified` is a subtype of
+      // the value type given by `type_hint`
+      if (!type_hint) {
+        dict_value->setType(DictType::create(k->type(), *unified));
```
Comment (gmagogsfm) [inline torch/csrc/jit/frontend/ir_emitter.cpp:1504]: For key type inference, is it:

1) check if there is type annotation, if so, check if all key types are subtypes of annotation. Then use annotation
2) if there is no type annotation, use key type of first pair in dictionary. 

Is this correct?

**Review by gmagogsfm — APPROVED** [review]

**Comment by ansley** [comment]
@ansley has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28231839).

**Comment by ansley** [comment]
@ansley has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28231839).

**Comment by ansley** [comment]
@ansley has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28231839).

**Comment by ansley** [comment]
@ansley has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28231839).

**Comment by ansley** [comment]
@ansley has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28231839).

**Comment by ansley** [comment]
@ansley has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28231839).

**Comment by ansley** [comment]
@ansley has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28231839).

**Comment by ansley** [comment]
@ansley has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28231839).
